shakenfist/blob.py

```diff
@@ -265,30 +262,24 @@ def ensure_local(self, locks):
         os.rename(blob_path + '.partial', blob_path)
         self.observe()
 
-    def request_replication(self):
-        targets = config.BLOB_REPLICATION_FACTOR - len(self.locations)
+    def request_replication(self, allow_excess=0):
+        replica_count = len(self.locations)
+        targets = config.BLOB_REPLICATION_FACTOR + allow_excess - replica_count
+        self.log.info('Desired replica count is %d, we have %d, excess of %d requested'
+                      % (config.BLOB_REPLICATION_FACTOR, replica_count, allow_excess))
         if targets > 0:
             blob_size_gb = int(int(self.size) / GiB)
-            by_disk = {}
-            for n in Nodes([node_active_states_filter]):
-                metrics = db.get_metrics(n.fqdn)
-
-                # If replicating a blob to a given node would make the target node
-                # have less than the minimum free disk, then don't do it.
-                disk_free_gb = int(metrics.get('disk_free', '0') / GiB)
-                if disk_free_gb - blob_size_gb < config.MINIMUM_FREE_DISK:
-                    continue
+            nodes = nodes_by_free_disk_descending(
+                minimum=blob_size_gb + config.MINIMUM_FREE_DISK)
+            if config.NODE_NAME in nodes:
+                nodes.remove(config.NODE_NAME)
 
-                by_disk.setdefault(disk_free_gb, [])
-                by_disk[disk_free_gb].append(n.fqdn)
-
-            for node in list(_sort_by_key(by_disk))[:targets]:
-                etcd.enqueue(node, {
+            for n in nodes[:targets]:
+                etcd.enqueue(n, {
                     'tasks': [FetchBlobTask(self.uuid)]
                 })
-                LOG.with_fields({
-                    'blob': self.uuid,
-                    'node': node
+                self.log.with_fields({
```

Comment:
I have kind of fallen into a rut where I always use `with_fields()`, to the extent that I have started thinking about removing `with_field()` and the related helpers such as `with_instance()`. I think it would be good to have a conversation about if these extra helpers add much, and if the simplicity is worth having fewer ways of expressing the same thing.

Reply:
`with_object()` automatically uses the correct object label. That is important for parameterised logging.

`with_instance()` et al could probably be removed in favour of `with_object()`